src/main/java/de/tum/in/www1/artemis/service/ParticipationVcsAccessTokenService.java

```diff
@@ -0,0 +1,145 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.DomainObject;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.participation.ParticipationVCSAccessToken;
+import de.tum.in.www1.artemis.domain.participation.ProgrammingExerciseStudentParticipation;
+import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
+import de.tum.in.www1.artemis.repository.ParticipationVCSAccessTokenRepository;
+import de.tum.in.www1.artemis.repository.ProgrammingExerciseStudentParticipationRepository;
+import de.tum.in.www1.artemis.service.connectors.localvc.LocalVCPersonalAccessTokenManagementService;
+
+@Profile(PROFILE_CORE)
+@Service
+public class ParticipationVcsAccessTokenService {
+
+    private static final Logger log = LoggerFactory.getLogger(ParticipationVcsAccessTokenService.class);
+
+    private final ParticipationVCSAccessTokenRepository participationVcsAccessTokenRepository;
+
+    private final ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository;
+
+    public ParticipationVcsAccessTokenService(ParticipationVCSAccessTokenRepository participationVCSAccessTokenRepository,
+            ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository) {
+        this.participationVcsAccessTokenRepository = participationVCSAccessTokenRepository;
+        this.programmingExerciseStudentParticipationRepository = programmingExerciseStudentParticipationRepository;
+    }
+
+    /**
+     * Creates a vcs access token for a User,Participation pair and stores it in the database
+     *
+     * @param user          the user which is owner of the token
+     * @param participation the participation which belongs to the token
+     * @return the newly created ParticipationVCSAccessToken
+     */
+    public ParticipationVCSAccessToken createParticipationVCSAccessToken(User user, StudentParticipation participation) {
+        ParticipationVCSAccessToken participationVCSAccessToken = new ParticipationVCSAccessToken();
+        participationVCSAccessToken.setUser(user);
+        participationVCSAccessToken.setParticipation(participation);
+        participationVCSAccessToken.setVcsAccessToken(LocalVCPersonalAccessTokenManagementService.generateSecureVCSAccessToken());
+        return participationVcsAccessTokenRepository.save(participationVCSAccessToken);
+    }
+
+    /**
+     * Creates vcs tokens for StudentParticipations which do not yet have one
+     *
+     * @param participations The participations, for which tokens should be generated, if they do not already have one
+     */
+    public void createMissingParticipationVCSAccessTokens(List<ProgrammingExerciseStudentParticipation> participations) {
```

Comment:
Is this method only meant for testing? At least I don't see any other place where it is called.

Reply:
It's used in `src/main/java/de/tum/in/www1/artemis/service/ParticipationVcsAccessTokenService.java:114`, isn't it?

Reply:
Yeah, it was only used by the test after removing the startup migration. Moved the code in a simplified version into the test